ga4mp/store.py

```diff
@@ -0,0 +1,125 @@
+    def __init__(self):
+        self = {
+            "user_properties": {},
+            "session_parameters": {}
+        }
+
+    def load(self):
+        raise NotImplementedError("Subclass should be using this function, but it was called through the base class instead.")
+
+    def save(self):
+        raise NotImplementedError("Subclass should be using this function, but it was called through the base class instead.")
+
+    def _check_exists(self, key):
+        # Helper function to make sure a key exists before trying to work with values within it.
+        if key not in self.keys():
+            self[key] = {}
+
+    def _set(self, param_type, name, value):
+        # Helper function to set a single parameter (user or session or other).
+        self._check_exists(key=param_type)
+        self[param_type][name] = value
+
+    def _get_one(self, param_type, name):
+        # Helper function to get a single parameter value (user or session).
+        self._check_exists(key=param_type)
+        return self[param_type].get(name, None)
+
+    def _get_all(self, param_type=None):
+        # Helper function to get all user or session parameters - or the entire dictionary if not specified.
+        if param_type is not None:
+            return self[param_type]
+        else:
+            return self
+
+    # While redundant, the following make sure the distinction between session and user items is easier for the end user.
+    def set_user_property(self, name, value):
+        self._set(param_type="user_properties", name=name, value=value)
+
+    def get_user_property(self, name):
+        return self._get_one(param_type="user_properties", name=name)
+
+    def get_all_user_properties(self):
+        return self._get_all(param_type="user_properties")
+
+    def clear_user_properties(self):
+        self["user_properties"] = {}
+
+    def set_session_parameter(self, name, value):
+        self._set(param_type="session_parameters", name=name, value=value)
+
+    def get_session_parameter(self, name):
+        return self._get_one(param_type="session_parameters", name=name)
+
+    def get_all_session_parameters(self):
+        return self._get_all(param_type="session_parameters")
+
+    def clear_session_parameters(self):
+        self["session_parameters"] = {}
+
+    # Similar functions for other items the user wants to store that don't fit the other two categories.
+    def set_other_parameter(self, name, value):
+        self._set(param_type="other", name=name, value=value)
+
+    def get_other_parameter(self, name):
+        return self._get_one(param_type="other", name=name)
+
+    def get_all_other_parameters(self):
+        return self._get_all(param_type="other")
+
+    def clear_other_parameters(self):
+        self["other"] = {}
+
+class DictStore(BaseStore):
+    # Class for working with dictionaries that persist for the life of the class.
+    def __init__(self):
+        super().__init__()
+
+    def load(self, data):
```

Comment:
This is cheating with Python :)
You are overwriting load function from the base class, but the load function in the base class does not have any parameters.
So you should not have any parameters either in the derived class.

Things DictStore would need, must be supplied in a constructor. So I would do:
```
def __init__(self, data: dict = None):
    super().__init__()
    if data:
        self.update(data)
```

Reply:
@zvizdo I updated the code as suggested, and it seems to work fine.

That said, I didn't make any changes to the `load()` function inside of DictStore - nor did I make any changes to the BaseStore class `__init__()` or `load()` functions.

So... are my changes sufficient or do I need to make more updates?

Reply:
After further testing and second guessing myself, I updated the BaseStore functions thusly:
```py
class BaseStore(dict):
    def __init__(self, data: dict = None, data_location: str = None):
        self.update([("user_properties", {}),("session_parameters", {})])

    def load(self, data: dict = None, data_location: str = None):
        raise NotImplementedError("Subclass should be using this function, but it was called through the base class instead.")

    def save(self, data_location: str = None):
        raise NotImplementedError("Subclass should be using this function, but it was called through the base class instead.")
```

It still appears to work just as it was, so let me know if that's wrong?

Reply:
@rptomey The key question to ask here is, what needs to happen if I want to write a MySQL Store (or anything else) to hold my stuff? 
MySQL needs IP address, username, password, and database to connect properly. All those are parameters to the MySQLStore class. But based on your current implementation I would need to go into the core code and add all that to BaseStore as well as params, even tho BaseStore doesn't need to know that.

Here would be the way:
```
class BaseStore(dict):
    def __init__(self):
        self.update([("user_properties", {}),("session_parameters", {})])

    def load(self):
        raise NotImplementedError()

    def save(self):
        raise NotImplementedError()

class DictStore(BaseStore):
    def __init__(self, data: dict = None):
        super()
        if data:
            self.update(data)  # in case you bring in data from somewhere else

    def load(self):
        ....

    def save(self):
        ....

class FileStore(BaseStore):
    def __init__(self, data_location: str = None):
        super()
        self._data_location = data_location if data_location else 'somedefaultpath.json'

    def load(self):
        .... # use self._data_location to load data

    def save(self):
        ....  # use self._data_location to save data

```

Reply:
That was helpful. Thanks! I've actually simplified things based on this feedback. I removed `load()` entirely from BaseStore and DictStore, and then in FileStore, I renamed it `_load_file()` to distinguish it from JSON's `load()` function and also to make it clear the user isn't supposed to call it.